[1/N] Introduce Mooncake Backend and Mooncake EP to Support Elastic EP

UNIDY2002 opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

<!-- Describe the purpose and goals of this pull request. -->

**Why do we need a new backend?**

The Mooncake Backend is a collective communication backend that provides fault tolerance while remaining seamlessly compatible with PyTorch. To the best of our knowledge, this is the first backend that supports fault tolerance for both:
- `torch.distributed` primitives, and
- EP all-to-all primitives.

This lays the foundation for fault-tolerant distributed MoE inference (#8961).

## Modifications

<!-- Detail the changes made in this pull request. -->

The Mooncake Backend and Mooncake EP can be enabled via two options:

**Mooncake Backend**
- Set `--dist-backend mooncake`.
- This replaces NCCL, Gloo, etc. used by `torch.distributed` with Mooncake.
- Implemented by registering `"mooncake"` as a custom backend in `torch.distributed.init_process_group`.
- Minimal code changes are required (non-intrusive integration).

**Mooncake EP**
- Set `--moe-a2a-backend mooncake`.
- This replaces the DeepEP all-to-all backend with Mooncake EP.
- Implemented by adding a new class `MooncakeEPDispatcher` in `token_dispatcher/mooncake.py`.

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

- Added a new unittest: `ep/test_mooncake_ep_small.py` to validate correctness of the new features.
- All existing unittests remain unaffected and should continue to pass.

**Caveat:**

The Mooncake package with the new backend support will be published soon.

For now, you may download the wheel named with "+ep" from the CI artifacts of https://github.com/kvcache-ai/Mooncake/pull/805.

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

According to our unit tests, enabling the Mooncake Backend and Mooncake EP may slightly decrease performance. This is expected, since the Mooncake Backend is not yet fully optimized; optimization work will be carried out separately.

We compared `test_mooncake_ep_small.py` with `test_deepep_small.py`.

**Throughput (token/s):**

| TestCase       | test\_mooncake\_ep\_small.py | test\_deepep\_small.py |
| -------------- | ---------------------------- | ---------------------- |
| TestPureDP     | 1581.135                     | 2011.321               |
| TestHybridDPTP | 1324.707                     | 1773.038               |
| TestTP         | 1484.145                     | 1668.714               |

**Latency (s):**

| TestCase       | test\_mooncake\_ep\_small.py | test\_deepep\_small.py |
| -------------- | ---------------------------- | ---------------------- |
| TestPureDP     | 11.561                       | 9.063                  |
| TestHybridDPTP | 13.639                       | 10.206                 |
| TestTP         | 12.122                       | 10.523                 |

## pd disaggregation use
We conducted an end-to-end test of this pr in a PD separation scenario on two H20-3e

**prefill**
```
GLOO_SOCKET_IFNAME=eth0 NCCL_IB_HCA=mlx5_ NCCL_IB_DISABLE=0 NCCL_SOCKET_IFNAME=eth0 CCL_IB_GID_INDEX=3 SGL_DISABLE_TP_MEMORY_INBALANCE_CHECK=1 NCCL_MIN_NCHANNELS=24 NCCL_IB_QPS_PER_CONNECTION=8 SGL_ENABLE_JIT_DEEPGEMM=1 python3 -m sglang.launch_server --disaggregation-ib-device  "mlx5_1,mlx5_2,mlx5_3,mlx5_4" --model-path /data/models/DeepSeek-R1-0528/ --tp 8 --disaggregation-mode prefill  --host 0.0.0.0 --port 30300 --moe-a2a-backend deepep --deepep-mode normal --disable-radix-cache --max-running-requests 16 --moe-dense-tp-size 1 --chunked-prefill-size 0 --trust-remote-code --watchdog-timeout 1000000  --enable-dp-attention --dp-size 4  --mem-fraction-static 0.8 --show-time-cost  --enable-dp-lm-head --page-size 64 --enable-eplb --eplb-rebalance-num-iterations 10000 --load-balance-method round_robin
```
**decode**

```
GLOO_SOCKET_IFNAME=eth0 NCCL_IB_DISABLE=0 NCCL_SOCKET_IFNAME=eth0 SGL_DISABLE_TP_MEMORY_INBALANCE_CHECK=1 NCCL_MIN_NCHANNELS=24 NCCL_IB_QPS_PER_CONNECTION=8  SGL_ENABLE_JIT_DEEPGEMM=1 SGL_DISABLE_TP_MEMORY_INBALANCE_CHECK=1 MC_CUSTOM_TOPO_JSON=/sgl-workspace/nic_priority_matrix.json MC_WORKERS_PER_CTX=1 python3 -m sglang.launch_server --model-path /data/models/DeepSeek-R1-0528/ --tp 8 --disaggregation-mode decode  --disaggregation-ib-device  "mlx5_1,mlx5_2,mlx5_3,mlx5_4" --host 0.0.0.0 --port 30300  --moe-a2a-backend mooncake   --disable-radix-cache --mem-fraction-static 0.8 --max-running-requests 1024  --moe-dense-tp-size 1 --cuda-graph-bs 32  --watchdog-timeout 1000000  --enable-dp-attention --dp-size 8 --trust-remote-code --show-time-cost --enable-dp-lm-head --nnodes 1 --node-rank 0  --enable-metrics --decode-log-interval 1 --enable-request-time-stats-logging --warmups 10 --log-level debug --log-requests --log-requests-level 3 --prefill-round-robin-balance --dist-backend mooncake
```

**router**
```
python3 -m sglang_router.launch_router --host 0.0.0.0 --port 8000 --pd-disaggregation --prefill-policy cache_aware --decode-policy round_robin --service-discovery --service-discovery-port 30300 --prefill-selector role=leader component=prefill --decode-selector role=leader component=decode --service-discovery-namespace default --request-timeout-secs 600 --max-concurrent-requests 1024 --model-path /data/models/DeepSeek-R1-0528/
```

**simpole curl**
```
curl http://127.0.0.1:8000/v1/chat/completions -H "Content-Type: application/json" -d '{"model": "/data/models/DeepSeek-R1-0528", "temperature": 0, "messages": [ {"role": "user", "content": "做一份北京出行攻略"}], "max_tokens": 300}'

{"id":"3f890c4a2b5445c0aabffdab05469787","object":"chat.completion","created":1757923786,"model":"/data/models/DeepSeek-R1-0528","choices":[{"index":0,"message":{"role":"assistant","content":"<think>\n嗯，用户想要一份北京出行攻略。这个需求很常见，但需要仔细考虑用户可能的背景和需求深度。用户没有说明旅行时长、预算、同行人员或兴趣偏好，所以攻略需要保持通用性，同时覆盖不同人群的可能需求。\n\n北京作为超大型旅游城市，攻略必须分层设计。首先考虑经典路线，这是大多数首次游客的核心需求。故宫、长城、天安门这些标志性景点必须包含，但要注意提醒用户提前预约——最近几年北京热门景点不预约根本进不去，很多外地游客会忽略这点。\n\n交通部分要特别强调地铁的便利性。北京地面交通太拥堵，地铁虽然拥挤但准时，尤其是1号线、2号线这些经典线路。不过也要提醒避开早晚高峰，不然用户体验会很差。机场快轨和北京南站到市区的方式也得写清楚，这是游客刚落地最需要的实用信息。\n\n美食推荐要平衡名气和本地特色。全聚德虽然被说商业化，但第一次来的人总想尝尝；同时必须加入尹三豆汁、门框卤煮这种地道小吃，但得标注清楚“豆汁慎点”——很多人喝不惯。四季民福烤鸭比全聚德性价比高，这个提示对预算有限的游客很实用。\n\n考虑到不同游客类型，补充了亲子、文化等主题路线。带孩子的家庭肯定需要环球影城和科技馆信息；文艺青年可能更喜欢798和胡同。还想到最近流行的","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"length","matched_stop":null}],"usage":{"prompt_tokens":8,"total_tokens":308,"completion_tokens":300,"prompt_tokens_details":null,"reasoning_tokens":0},"metadata":{"weight_version":"default"}}
```


## Next Steps

This PR establishes the foundation by introducing the Mooncake Backend and Mooncake EP. The following steps will build upon it:
- **Integrate with EPLB**: Adapt the EPLB algorithm to avoid mapping experts to faulty ranks, using Mooncake’s fault-awareness.
- **Expert Redistribution**: Trigger expert rebalancing when new rank failures occur, ensuring that every expert has at least one healthy replica.
- **Scheduler Enhancements**: Make the scheduler fault-aware, so failed requests can be retried or rerouted to healthy ranks automatically.
- **Stability and Performance Testing**: Broaden testing coverage with larger TP sizes, prefill-decode disaggregation, and multiple-rank failure scenarios.

These will be addressed in subsequent PRs as outlined in the milestone roadmap in https://github.com/sgl-project/sglang/pull/8961#issuecomment-3269511514.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

UNIDY2002 committed:
Introduce Mooncake Backend and Mooncake EP

HanHan009527 committed:
tiny fix mooncake pr (#12)

* fix

* fix

* tiny fix

* fix

zhyncs committed:
Merge branch 'main' into mooncake-pr

UNIDY2002 committed:
Fix for more readable code

UNIDY2002 committed:
Merge branch 'main' into mooncake-pr

whybeyoung commented:
LGTM

HanHan009527 committed:
lint

UNIDY2002 committed:
API: change broken_ranks to active_ranks

ShangmingCai commented on python/sglang/srt/layers/moe/token_dispatcher/mooncake.py:
```diff
@@ -0,0 +1,392 @@
+        return DispatchOutputFormat.DEEPEP_LL
+
+
+assert isinstance(MooncakeDispatchOutput, DispatchOutput)
```
```suggestion
assert issubclass(MooncakeDispatchOutput, DispatchOutput)
```

ch-wan replied:
I once tried `issubclass` but the engine crashes.

ShangmingCai commented on python/sglang/srt/layers/moe/token_dispatcher/mooncake.py:
```diff
@@ -0,0 +1,392 @@
+        return CombineInputFormat.DEEPEP_LL
+
+
+assert isinstance(MooncakeCombineInput, CombineInput)
```
```suggestion
assert issubclass(MooncakeCombineInput, CombineInput)
```

UNIDY2002 committed:
Fix issues according to review

UNIDY2002 committed:
Fix an issue

UNIDY2002 commented on python/sglang/srt/layers/moe/token_dispatcher/mooncake.py:
```diff
@@ -0,0 +1,392 @@
+_ACTIVE_RANKS: Optional[torch.Tensor] = None
+
+
+def get_ep_active_ranks() -> torch.Tensor:
+    assert _ACTIVE_RANKS is not None, "_ACTIVE_RANKS is not initialized"
+    return _ACTIVE_RANKS
```
Yes. In the following PRs, `_ACTIVE_RANKS` will be used by the model_runner to know whether a rank has been broken and determine whether to trigger a redistribution of expert weights.

UNIDY2002 committed:
Merge branch 'main' into mooncake-pr

UNIDY2002 committed:
Bug-fix

UNIDY2002 committed:
Bug-fix

ch-wan commented on python/sglang/srt/models/deepseek_v2.py:
```diff
@@ -591,7 +592,7 @@ def __init__(
         self.top_k = config.num_experts_per_tok
 
-        if get_moe_a2a_backend().is_deepep():
+        if not get_moe_a2a_backend().is_none():
```
I prefer `is_deepep` or `is_mooncake` here as we will support other a2a backends.

ch-wan commented on python/sglang/srt/models/deepseek_v2.py:
```diff
@@ -620,7 +621,7 @@ def __init__(
             )
 
-        self._enable_deepep_moe = get_moe_a2a_backend().is_deepep()
+        self._enable_a2a_moe = not get_moe_a2a_backend().is_none()
```
same here

ch-wan commented on python/sglang/srt/server_args.py:
```diff
@@ -1404,6 +1411,11 @@ def add_cli_args(parser: argparse.ArgumentParser):
             help="The device to use ('cuda', 'xpu', 'hpu', 'npu', 'cpu'). Defaults to auto-detection if not specified.",
         )
+        parser.add_argument(
+            "--enable-mooncake-dist-backend",
```
how about adding `--elastic-ep-backend` with one option `mooncake`?

ch-wan commented on python/sglang/srt/server_args.py:
```diff
@@ -190,6 +190,7 @@ class ServerArgs:
 
     # Runtime options
     device: Optional[str] = None
+    enable_mooncake_dist_backend: bool = False
```
please update `server_arguments.md` whenever you modified args list.

UNIDY2002 committed:
Change arg `--enable-mooncake-dist-backend` to `--elastic-ep-backend mooncake`

UNIDY2002 committed:
Fix format

UNIDY2002 committed:
Merge branch 'main' into mooncake-pr

UNIDY2002 committed:
Add ep/test_mooncake_ep_small.py to test suite, and a couple of fixes:

- Add an argument `--mooncake-ib-device` to allow specifying the IB devices to use
- Keep in line with the upstream API changes of token_dispatcher
- Still use deepep as the `--moe-a2a-backend` in CI because of some unfulfilled MMIO permission requirements

UNIDY2002 committed:
Merge branch 'main' into mooncake-pr

ShangmingCai reviewed (approved):
LGTM.

UNIDY2002 committed:
Remove '--mooncake-ib-device' from tests

Removed '--mooncake-ib-device' and associated device list from multiple test cases.

UNIDY2002 committed:
Merge branch 'main' into mooncake-pr

UNIDY2002 committed:
Revert "Remove '--mooncake-ib-device' from tests"

This reverts commit 3bc4150b2aec45095467cc2994491716273d8b74.

UNIDY2002 committed:
Merge branch 'main' into mooncake-pr

UNIDY2002 committed:
Merge branch 'main' into mooncake-pr

ShangmingCai committed:
Merge branch 'main' into mooncake-pr

ShangmingCai committed:
upd

ShangmingCai committed:
Update run_suite.py

UNIDY2002 committed:
Install Mooncake+EP in `ci_install_deepep.sh` script

ShangmingCai committed:
fix ci_install_deepep.sh

UNIDY2002 committed:
Merge branch 'main' into mooncake-pr

UNIDY2002 committed:
Merge branch 'main' into mooncake-pr

UNIDY2002 committed:
Fix format

ShangmingCai committed:
Merge branch 'main' into mooncake-pr
